### ISSUE TITLE
Port Powershell based Hyper-V API to WMI

### ISSUE THREAD

**Issue by jefflill** [nforgeio/neonSDK#38]
We're currently using Powershell Hyper-V cmdlets to manage Hyper-V.  This is slow and clunky and it looks like there are alternative APIs:

https://learn.microsoft.com/en-us/windows/win32/hyperv_v2/windows-virtualization-portal
https://learn.microsoft.com/en-us/previous-versions/windows/desktop/virtual/using-the-virtualization-wmi-provider

and here are some Hyper-V WMI samples:

https://github.com/microsoft/Windows-classic-samples/tree/main/Samples/Hyper-V

Here's an overview of WMI for .NET:

https://learn.microsoft.com/en-us/archive/msdn-magazine/2002/may/wmi-and-net-system-management-lets-you-take-advantage-of-wmi-apis-within-managed-code

**Comment by jefflill** [comment]
Here's the inventory of operations we'll need to port:

- [ ] Resize-VHD
- [ ] SetVM-Name
- [ ] New-VHD
- [ ] Add-VMHardDisk
- [ ] Set-VM-CheckpointType
- [ ] Set-VMProcessor
- [ ] Set-VMNetworkAdapter
- [ ] Remove-VM-Name
- [ ] Get-VM
- [ ] Start-VM
- [ ] Save-VM
- [ ] Mount-VHD
- [ ] Dismount-VHD
- [ ] Optimize-VHD
- [ ] Add-VMDvdDrive
- [ ] Remove-VMDvdDrive
- [ ] Get-NetAdapter
- [ ] New-VMSwitch
- [ ] New-NetIPAddress
- [ ] New-NetNAT
- [ ] Remove-VMSwitch
- [ ] Remove-NetNAT
- [ ] Get-VMNetworkAdapter
- [ ] Get-NetNAT
- [ ] Get-NetIPAddress

**Comment by jefflill** [comment]
I'm starting to learn my way around WMI and it looks like things are a bit more complicated at the WMI level as opposed to the PowerShell cmdlets.  For example, Hyper-V switches are abstracted by PowerShell to have switch types (External, Internal, and Private) where as this is not a intrinsic property exposed by WMI switches.

It looks like the PowerShell and Hyper-V Manager UI also handles things like managing switch ports and their connections to VM network adpators, which we'd need to do as well.

All of this is possible, but this is not really a priority right now; we need to ship neonKUBE and the PowerShell approach has been working since 2016.

I'm puting this back into **deep-freeze**.

**Comment by jefflill** [comment]
This guy has figured this out:

https://github.com/jscarle/HyperV.NET

by using **ILSpy** to decompile the Hyper-V cmdlets:

https://learn.microsoft.com/en-us/archive/blogs/luisdem/get-the-source-code-of-the-powershell-cmdlets

So this task should be much easier now, since I just need to replicate what a few cmdlets are doing.

I'm moving this from **deep-freeze** to the **public release**.

**Comment by jefflill** [comment]
I've reopened this because this is really impacting **neon-desktop** CPU utilization and it's just bad to have include PowerShell in the installer, which bloats the installer size by 100MB and the on-disk footprint by 300MB.

**Comment by jefflill** [comment]
**DONE:** This was a real head-banger but it's finally working now.

I ended up using ILSpy to decompile the Hyper-V related cmdlets, modify them a tiny bit, and then include them in the **Neon.HyperV** library, calling them directly using PowerShell runspaces.

Other than losing perhaps a week to COVID, I encountered what I thought was one problem that were actually a handful of unrelated problems (which was confusing).

* I initially installed and used the latest stable version of **ILSpy**, but it clipped file name segments to just 30 characters which effectively caused some of the ILSpy decompled files to be overwritten.  It turns out that the latest pre-release ILSpy release fixed that, but it took me a day or to to figure out.

* I also decompiled the .NET Framework version of the Hyper-V cmdlets (for PowerShell 5.0) and that resulted in some compilation and other conflicts.  This took another day or so to figure out.

* It looked like basic Hyper-V cmdlets were working in new neonSDK unit tests, but when I ran the VM tests 100 times, I saw that maybe 1 out of 5-6 tests were hanging.  I thought that I wasn't invoking the cmdlets correctly, but after a couple of days, I realized that there were other problems:

  * Virtual machines aren't always ready to accept a command (e.g. when a previous command is still in progress).  I fixed this by spinning until the VM reports being ready.  This can result in the VM getting stuck in a weird state where it doesn't respond to PowerShell command or even the Hyper-V Manager UI, leaving killing the VM process the only way to move on.
  * Virtual machine operations don't always complete synchronously.  The fix was to wait for the operations to complete in my code.
  * Graceful VM shutdown didn't work sometimes, leaving the VM running and the unit test hanging due to waiting for the VM to stop, which never happened.  The unit test case starts and then nearly immediately stops the VM.  I believe the problem here is due to the guest integrations services on the VM not having the chance to start before the unit test sends the shutdown signal, resulting in the signal being lost.  The workaround here was to continue periodically shutting down the VM while waiting for it to actually stop.  This assumes that sending multiple shutdown signals won't cause a problem (which seems to be the case).
  * One unit test inserted and then immediately ejected a DVD ISO file.  This would also hang.  The workaround here was to record the time when the DVD was inserted and then before commanding the eject, I'd compare that current time to insert time and then pause long enough for at least 5 seconds has elapsed between the insert and the eject.  This seems like a Hyper-V issue.

I believe a lot of these problems were also possible with the old PowerShell implementation, but that the slowness of that approach masked these problems.
